data/libs/SpaceStation.lua

```diff
@@ -291,8 +291,10 @@ end
 
 -- add num random ships for sale to station ShipMarket
 local function addRandomShipAdvert(station, num)
+	local playersShip = Game.player:GetShipType()
 	for i=1,num do
-		local avail = station.type == "STARPORT_SURFACE" and groundShips or spaceShips
+		local avail_0 = station.type == "STARPORT_SURFACE" and groundShips or spaceShips
+		local avail = utils.build_array(utils.filter(function (k,def) return (playersShip == def.name or Calibration.prerequisite(playersShip,def.name)) end, pairs(avail_0)))
```

Comment:
I'm not too sure about this feature. Do we really want to lock players into a "ship track"? I don't think it makes much sense from an in-universe point of view, it's as if you went to your auto dealer, and he'd refuse to sell you a mid-range Toyota because you drive a Volkswagen

Reply:
I agree with that. When I'm thinking about new ships, and ship lines from manufacturers, I expressedly want to avoid providing a clear progression through ships. I'd like quite the opposite really. Player choices based on multiple factors, and no clear better ship for every situation

Reply:
**It gives more control to developers on the progress of game and makes all ships useful.**
In current game, in large degree, and in this PR, not so much, considerable number of ships is not strategically correct to be used, without this feature. _(I am sad for their designers.)_ When i thought about in-universe point of view, my thinking was that we speak about spaceships not cars. Specific experience is needed for each model. Dealer will say 'Sorry you have not the experience to drive Volkswagen-IV. You may start by Volkswagen-I if you like or, since you are experienced pilot  on Toyota-VI, i can sell you a new one like yours or Toyota-VII if you like. That's the law.'
We can disable the feature by setting default tables empty. If any ship builder wants he can fill prerequisites for his ships. That is why tag in json files is better place for it.
The game is surely playable without it, but some ships will not be bought by best player's strategy. I played both and prefer the feature.

Reply:
But that would add bureaucracy in the game, it feels like. Why wouldn't you be able to train for this and this ship directly for example? Or size classifications for example. 
Locking the player into lines of progression feels rather limiting to me, if I understand this correctly.
True, there are ships that aren't really useful in a sense, but this direction to fix that feels wrong.  It would be in my opinion more meaningful to provide more ways to distinguish between them. Like decoupling mass and ship health/armor, or the 1t cargo capacity unit in favor of m3. Or to introduce actual equipment slots and such, instead of a flat volume to hold everything together. So the ship designer can actually provide a fighter ship which isn't just a little transport that happens to have guns for example.

Reply:
Forgot to mention that another approach, more free, was to use the equilibrium equations, i describe in my article. In that approach ship prices drops as player progress to larger ships. I was not totally sure  if the duration from ship to next ship could stay steady and found hard to find 'rn's constants that would not disturb current state too much. As lua code, could be easily applied but since ship prices are handled by C, I found it too messy to ask to define them dynamically or to replace them in a similar way i did with prerequisite feature.